Ban manual implementation of TryFuture and TryStream

`Future<Output = Result<T, E>>` automatically implements `Tryfuture`. Also it can call `.await` and `FutureExt`'s methods without `TryFuture::into_future`.

Also, we may want to rewrite this trait to inherit `Future` in the future, so I think it makes sense to ban manual implementation for forward compatibility.
https://github.com/rust-lang-nursery/futures-rs/blob/78c5d852ec24edcf630a4c569c1234d078416f4f/futures-core/src/future/mod.rs#L61-L63

I'm thinking of using [the approach that `SliceIndex` uses](https://github.com/rust-lang/rust/blob/e1d7e4ae82c5500407e606c0b2b595597b9981f3/src/libcore/slice/mod.rs#L2621), but there may be better ways.

Yeah, @aturon and I hit a bug w/ cycles in the trait checker when attempting to do that:

```rust
trait Future { 
    type Output;
}

trait TryFuture: Future<Output = Result<Self::Ok, Self::Err>> {
    type Ok;
    type Err;
}
```
gives
```
error[E0391]: cycle detected when computing the supertraits of `TryFuture`
 --> src/main.rs:5:1
  |
5 | trait TryFuture: Future<Output = Result<Self::Ok, Self::Err>> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: ...which again requires computing the supertraits of `TryFuture`, completing the cycle
note: cycle used when collecting item types in top-level module
 --> src/main.rs:5:1
  |
5 | trait TryFuture: Future<Output = Result<Self::Ok, Self::Err>> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```
IIRC @nikomatsakis said this was fixed in chalk, but there's no ETA for the fix as far as I know.

--- a/futures-core/src/future/mod.rs
+++ b/futures-core/src/future/mod.rs
@@ -47,9 +47,17 @@ where
     }
 }
 
+mod private_try_future {
+    use super::Future;
+
+    pub trait Sealed {}
+
+    impl<F, T, E> Sealed for F where F: ?Sized + Future<Output = Result<T, E>> {}
+}
+
 /// A convenience for futures that return `Result` values that includes
 /// a variety of adapters tailored to such futures.
-pub trait TryFuture {
+pub trait TryFuture: private_try_future::Sealed {
     /// The type of successful values yielded by this future
     type Ok;
 

--- a/futures-core/src/stream.rs
+++ b/futures-core/src/stream.rs
@@ -112,9 +112,17 @@ where
     }
 }
 
+mod private_try_stream {
+    use super::Stream;
+
+    pub trait Sealed {}
+
+    impl<S, T, E> Sealed for S where S: ?Sized + Stream<Item = Result<T, E>> {}
+}
+
 /// A convenience for streams that return `Result` values that includes
 /// a variety of adapters tailored to such futures.
-pub trait TryStream {
+pub trait TryStream: private_try_stream::Sealed {
     /// The type of successful values yielded by this future
     type Ok;
 

